File: expensereport-java/src/main/java/com/nelkinda/training/ExpenseType.java
package com.nelkinda.training;

public abstract class ExpenseType {
    public abstract String getName();
    public abstract boolean isMeal();
    public abstract boolean exceedsLimit(int amount);

    public static final ExpenseType DINNER = new ExpenseType() {
        @Override
        public String getName() {
            return "dinner";
        }

        @Override
        public boolean isMeal() {
            return true;
        }

        @Override
        public boolean exceedsLimit(int amount) {
            return amount > 5000;
        }
    };

    public static final ExpenseType BREAKFAST = new ExpenseType() {
        @Override
        public String getName() {
            return "breakfast";
        }

        @Override
        public boolean isMeal() {
            return true;
        }

        @Override
        public boolean exceedsLimit(int amount) {
            return amount > 1000;
        }
    };

    public static final ExpenseType CAR_RENTAL = new ExpenseType() {
        @Override
        public String getName() {
            return "car rental";
        }

        @Override
        public boolean isMeal() {
            return false;
        }

        @Override
        public boolean exceedsLimit(int amount) {
            return false; // No limit for car rental
        }
    };
}

File: expensereport-cxx/ExpenseType.cpp
#include "ExpenseType.h"

#include <climits>

const ExpenseType& ExpenseType::BREAKFAST  = ExpenseType("Breakfast",     1000, true);
const ExpenseType& ExpenseType::DINNER     = ExpenseType("Dinner",        5000, true);
const ExpenseType& ExpenseType::CAR_RENTAL = ExpenseType("Car Rental", INT_MAX, false);
const ExpenseType& ExpenseType::LUNCH      = ExpenseType("Lunch",         2000, true);

ExpenseType::ExpenseType(const std::string& name, const int limit, const bool isMeal) : name(name), limit(limit), isMeal(isMeal) {}
